### ISSUE TITLE
diarizer.py

### ISSUE THREAD

**Issue by runammm** [runammm/HabitLink#4]
**Labels:** `feature`, `audio-processing`, `module`

#### 1. Overview
To accurately analyze a user's speech habits in a conversational setting, we must first distinguish the user's speech from that of other participants. The project proposal requires the system to "extract only the user's voice" and consider the full conversation context for features like "Contextual Appropriateness Evaluation".

This task is to create a speaker diarization module that processes an audio recording and assigns a unique speaker label to each transcribed segment.

#### 2. Proposed Implementation
We will use the **WhisperX** library. It extends OpenAI's Whisper with highly accurate word-level timestamps and speaker diarization capabilities.

**Why WhisperX?**
* **Synergy:** It builds directly on the Whisper model you are already using.
* **All-in-One:** It combines transcription and diarization into a single, streamlined process, automatically aligning speaker labels with the transcribed text.
* **Performance:** It is a well-regarded, high-performance tool for this task.

This implementation will involve creating a new, dedicated module file: `src/diarizer.py`.

#### 3. Module Design (`src/diarizer.py`)
We will create a `SpeakerDiarizer` class to encapsulate the diarization logic.

```python
# src/diarizer.py

import whisperx
import torch

class SpeakerDiarizer:
    """
    A class to perform speaker diarization and transcription using WhisperX.
    """
    def __init__(self, device: str = "cpu", batch_size: int = 16, compute_type: str = "float32"):
        """
        Initializes the SpeakerDiarizer and loads the required models.

        Args:
            device (str): Device to run the models on ("cuda" or "cpu").
            batch_size (int): Batch size for transcription.
            compute_type (str): The compute type for the model (e.g., "int8", "float16", "float32").
        """
        self.device = device
        self.batch_size = batch_size
        self.compute_type = compute_type
        # Note: You can specify the whisper model size here, e.g., "large-v2"
        self.asr_model = whisperx.load_model("base", self.device, compute_type=self.compute_type, language="ko")
        self.diarize_model = whisperx.DiarizationPipeline(use_auth_token="YOUR_HF_TOKEN_HERE", device=self.device)

    def process(self, audio_path: str) -> list[dict]:
        """
        Transcribes and diarizes an audio file.

        Args:
            audio_path (str): The path to the input audio file.

        Returns:
            list[dict]: A list of transcript segments, each with 'start', 'end', 'text', and 'speaker' keys.
                        Example: [{'start': 0.5, 'end': 2.3, 'text': 'Hello there.', 'speaker': 'SPEAKER_00'}]
        """
        # 1. Load Audio
        audio = whisperx.load_audio(audio_path)

        # 2. Transcribe
        result = self.asr_model.transcribe(audio, batch_size=self.batch_size)

        # 3. Align Whisper output
        model_a, metadata = whisperx.load_align_model(language_code=result["language"], device=self.device)
        result = whisperx.align(result["segments"], model_a, metadata, audio, self.device, return_char_alignments=False)

        # 4. Diarize
        # Note: You can add min_speakers and max_speakers parameters if known.
        diarize_segments = self.diarize_model(audio)
        result = whisperx.assign_word_speakers(diarize_segments, result)

        # Format output for clarity
        formatted_segments = []
        for segment in result["segments"]:
            formatted_segments.append({
                "start": segment.get("start"),
                "end": segment.get("end"),
                "text": segment.get("text", "").strip(),
                "speaker": segment.get("speaker")
            })
        
        return formatted_segments
```

#### 4. Integration Plan
The new `SpeakerDiarizer` will be integrated into the `AudioInterface` class. The existing `record_and_transcribe` method can be updated or a new method can be created to handle diarization.

**`src/audio_interface.py` (Proposed Changes):**

```python
# src/audio_interface.py (Conceptual change)
from .audio_engine import AudioEngine
from .diarizer import SpeakerDiarizer # <-- Import the new class

class AudioInterface:
    def __init__(self, audio_engine: AudioEngine, diarizer: SpeakerDiarizer, language: str = "ko"):
        self.audio_engine = audio_engine
        self.diarizer = diarizer
        self.language = language # Language might be handled by the diarizer itself

    def record(self, duration: float):
        return self.audio_engine.record(duration)

    def record_and_process(self, duration: float) -> list[dict]:
        """
        Records audio for a given duration and returns the diarized transcript.
        """
        input_path = self.record(duration)
        diarized_transcript = self.diarizer.process(input_path)
        
        # Future step: Add logic here to filter for the primary user's speech
        # [cite_start]based on voice registration, as mentioned in the proposal[cite: 300, 301].
        # For now, we can return the full diarized transcript.

        return diarized_transcript
```

#### 5. Action Items
- [ ] Install necessary libraries: `pip install git+https://github.com/m-bain/whisperx.git`
- [ ] Update `requirements.txt` with `whisperx` and its dependencies (e.g., `torch`, `torchaudio`).
- [ ] Create a Hugging Face account and accept the user conditions for the `pyannote/speaker-diarization-3.1` model to get a User Access Token.
- [ ] Implement the `SpeakerDiarizer` class in a new file `src/diarizer.py`.
- [ ] Update `src/audio_interface.py` to use the `SpeakerDiarizer` class.
- [ ] Create a test script in `if __name__ == "__main__"` of `diarizer.py` to verify its functionality using a pre-recorded multi-speaker audio file.
